pkg/pgmodel/ingestor/series_writer.go

```diff
@@ -0,0 +1,358 @@
+// This file and its contents are licensed under the Apache License 2.0.
+// Please see the included NOTICE for copyright information and
+// LICENSE for a copy of the license.
+
+package ingestor
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/jackc/pgtype"
+	"github.com/timescale/promscale/pkg/pgmodel/common/schema"
+	"github.com/timescale/promscale/pkg/pgmodel/model"
+	"github.com/timescale/promscale/pkg/pgmodel/model/pgutf8str"
+	"github.com/timescale/promscale/pkg/pgxconn"
+)
+
+const (
+	seriesInsertSQL = "SELECT (_prom_catalog.get_or_create_series_id_for_label_array($1, l.elem)).series_id, l.nr FROM unnest($2::prom_api.label_array[]) WITH ORDINALITY l(elem, nr) ORDER BY l.elem"
+)
+
+type seriesWriter struct {
+	conn          pgxconn.PgxConn
+	labelArrayOID uint32
+}
+
+type labelInfo struct {
+	labelID int32
+	Pos     int32
+}
+
+type labelKey struct {
+	MetricName, Name, Value string
+}
+
+type SeriesVisitor interface {
+	VisitSeries(func(s *model.Series) error) error
+	NumSeries() int
+}
+
+func labelArrayTranscoder() pgtype.ValueTranscoder { return &pgtype.Int4Array{} }
+
+func NewSeriesWriter(conn pgxconn.PgxConn, labelArrayOID uint32) *seriesWriter {
+	return &seriesWriter{conn, labelArrayOID}
+}
+
+type perMetricInfo struct {
+	metricName             string
+	series                 []*model.Series
+	labelList              *model.LabelList
+	maxPos                 int
+	labelArraySet          *pgtype.ArrayType
+	labelArraySetNumLabels int
+}
+
+// Set all seriesIds for a samples, fetching any missing ones from the DB,
+// and repopulating the cache accordingly.
+func (h *seriesWriter) WriteSeries(sv SeriesVisitor) error {
+	infos := make(map[string]*perMetricInfo)
+	seriesCount := 0
+	err := sv.VisitSeries(func(series *model.Series) error {
+		if !series.IsSeriesIDSet() {
+			metricName := series.MetricName()
+			info, ok := infos[metricName]
+			if !ok {
+				info = &perMetricInfo{metricName: metricName, labelList: model.NewLabelList(10)}
+				infos[metricName] = info
+			}
+			info.series = append(info.series, series)
+			seriesCount++
+		}
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+	if len(infos) == 0 {
+		return nil
+	}
+
+	labelMap := make(map[labelKey]labelInfo, seriesCount)
+	//logically should be a separate function but we want
+	//to prevent labelMap from escaping, so keeping inline.
+	{
```

Comment:
I know we do like this in other places, but I really don't like it. I will try to do some refactoring in ingestor.

Reply:
This is purely because of the comment above. It should be a function but for performance it's not. This code is not new in the PR.